katdal/datasources.py

```diff
@@ -623,42 +622,49 @@ def from_url(cls, url, chunk_store='auto', upgrade_flags=True, **kwargs):
+        u = urllib.parse.urlparse(url, scheme='file')
         # Merge key-value pairs from URL query with keyword arguments
         # of function (the latter takes precedence)
-        url_kwargs = dict(urllib.parse.parse_qsl(url_parts.query))
+        url_kwargs = dict(urllib.parse.parse_qsl(u.query))
         url_kwargs.update(kwargs)
         kwargs = url_kwargs
         # Extract Redis database number if provided
         db = int(kwargs.pop('db', '0'))
-        if url_parts.scheme == 'file':
+        if u.scheme == 'file':
             # RDB dump file
-            telstate = katsdptelstate.TelescopeState()
+            telstate = katsdptelstate.TelescopeState(katsdptelstate.memory.MemoryBackend())
             try:
-                telstate.load_from_file(url_parts.path)
-            except OSError as err:
-                raise DataSourceNotFound(str(err))
-        elif url_parts.scheme == 'redis':
+                telstate.load_from_file(u.path)
+            except OSError as e:
+                raise DataSourceNotFound(str(e))
+        elif u.scheme == 'redis':
             # Redis server
             try:
-                telstate = katsdptelstate.TelescopeState(url_parts.netloc, db)
+                telstate = katsdptelstate.TelescopeState(u.netloc, db)
             except katsdptelstate.ConnectionError as e:
                 raise DataSourceNotFound(str(e))
-        elif url_parts.scheme in {'http', 'https'}:
-            # Open endpoint as an S3 object store (with auth info in kwargs)
-            endpoint = urllib.parse.urlunparse(url_parts[:2] + 4 * ('',))
-            rdb_store = S3ChunkStore.from_url(endpoint, **kwargs)
-            rdb_path = urllib.parse.urlunparse(url_parts[:3] + 3 * ('',))
-            telstate = katsdptelstate.TelescopeState()
-            with rdb_store.request('RDB', 'GET', rdb_path) as response:
-                telstate.load_from_file(io.BytesIO(response.content))
+        elif u.scheme in {'http', 'https'}:
+            # Treat URL prefix as an S3 object store (with auth info in kwargs)
+            store_path = os.path.dirname(os.path.dirname(u.path))
+            store_url = (u.scheme, u.netloc, store_path, '', '', '')
+            store_url = urllib.parse.urlunparse(store_url)
+            # Strip off parameters, query strings and fragments to get basic URL
+            rdb_url = (u.scheme, u.netloc, u.path, '', '', '')
+            rdb_url = urllib.parse.urlunparse(rdb_url)
+            telstate = katsdptelstate.TelescopeState(katsdptelstate.memory.MemoryBackend())
+            try:
+                rdb_store = S3ChunkStore.from_url(store_url, **kwargs)
+                with rdb_store.request('', 'GET', rdb_url) as response:
+                    telstate.load_from_file(io.BytesIO(response.content))
+            except StoreUnavailable as e:
+                raise DataSourceNotFound(str(e))
             if chunk_store == 'auto' and not kwargs.get('s3_endpoint_url'):
```

Comment:
There may be uses for disabling this behaviour. For example, if one wants to rerun the continuum imager by hand, you'll want to use the RDB file from the archive, but the data will be in the on-site Ceph store. Maybe have another magic value like 'original' ignore this bit of code and apply `infer_chunk_store`.

Reply:
Would you not have to specify `s3_endpoint_url` to point to the on-site Ceph store in this case?

Reply:
No, it's in the stream-specific `s3_endpoint_url` in telstate (the on-site versions of the data are given a different stream_name to disambiguate them).

Reply:
In that case I'll just refrain from inferring the chunk store too early. I thought it would save some time but the store opens quite quickly anyway.